omigami/spectra_matching/spectra_matching.py

```diff
@@ -183,27 +187,42 @@ def _make_batch_requests(
         parameters: SpectraMatchingParameters,
         endpoint: str,
     ) -> List[pd.DataFrame]:
+        self._failed_spectra = []
         batch = []
-        requests = []
+        predictions = []
 
         for spectrum in spectra_generator:
+            if self._get_cache_id(spectrum) in self._cached_results:
+                predictions.append(self._cached_results[self._get_cache_id(spectrum)])
+                continue
+
             batch.append(spectrum)
             if len(batch) == SPECTRA_LIMIT_PER_REQUEST:
-                payload = self._build_payload(batch, parameters)
-                requests.append(self._send_request(payload, endpoint))
+                response = self._send_request(batch, endpoint, parameters)
+                if response is not None:
+                    formatted_results = self._format_results(response)
+                    self._cache_results(formatted_results, batch)
+                    predictions.extend(formatted_results)
+
                 batch = []
 
         if batch:
-            payload = self._build_payload(batch, parameters)
-            requests.append(self._send_request(payload, endpoint))
+            response = self._send_request(batch, endpoint, parameters)
+            if response is not None:
+                formatted_results = self._format_results(response)
+                self._cache_results(formatted_results, batch)
+                predictions.extend(formatted_results)
```

Comment:
can you try refactoring that into a function nested inside this function and reuse it above to see if its better?

Reply:
I used your idea in my own way and it looks nicer. Testing is simpler now too which is a good sign.